### ISSUE TITLE
Feature: Resource creation dialog

### ISSUE THREAD

**Issue by adaburrows** [lightningrodlabs/rea-playspace#84]
For a new resource if one is created:
* name
* trackingIdentifier (let's use that for lot too if they need it)
* note
* primaryAccountable (set to the event receiver)
* stage (set to the process.basedOn ProcessSpecification if there was one)
* (maybe skip containedIn for now, which also means  we wouldn't be supporting `combine` and `separate`, could discuss)

**Comment by fosterlynn** [comment]
Some further notes:

Here is the resource page in VF: https://www.valueflo.ws/concepts/resources/.  It is pretty conceptual, possibly not super practical for this, unless you want more conceptualization.

Commitments can't create or update resources, only events.  Commitments can reference an existing inventoried resource for the situations where it matters which exact resource, like: use a specific machine; sell a used bike or whatever, it has exact traits and history unlike any other bike.  (Maybe not priority for Playspace, but fine if it does.)

Only certain actions can involve creation of a resource by an event, see [https://lab.allmende.io/valueflows/valueflows/-/merge_requests/706 in process.](https://lab.allmende.io/valueflows/valueflows/-/merge_requests/706#note_39933).  Only events can create resources.  Saving an event that creates a resource should ideally treat both as a "unit of work", i.e. if both don't get saved together, the other is backed out, and the save is an error.

I'm thinking we are not supporting `raise`, `lower`, `move` right now, as they are different user stories probably, is that other people's understanding?  

In the case of the UI where "Done" is what saves, and there might be multiple events, it would be important to save them in date/time order, so the last event's resource effects is what shows up.  (Quantities will be fine, because they will add up in the end, but things like primaryAccountable would need to be the last one.)

Here's the part of the graphql spec that applies, once we are on hREA, but may have general applicability too: https://lab.allmende.io/valueflows/vf-schemas/vf-graphql/-/blob/sprout/lib/schemas/observation.gql#L257.  And maybe some other nearby pieces referenced there.

**Comment by adaburrows** [comment]
I should be able to finish this up when I wake up later today (Friday). I got things to a point where it feels manageable:
* I want to summarize what will happen with the current EconomicEvent data.
* When a resource is created through a transfer for the first time, two EconomicResources need to be created: the resourceInventoriedAs and toResourceInventoriedAs
  * resourceInventoriedAs is credited (by both accountingEffect and onhandEffect)
  * toResourceInventoriedAs is debited (by both accountingEffect and onhandEffect)
* When something is not being transferred, then the resourceInventoriedAs contains the reference to the object being adjusted (by both accountingEffect and onhandEffect).
* If resources need to be created, then I want to give the operator (better than user, no?) the chance to customize the EconomicResources' fields that can be customized.
  * The fields the can be changed are:
    * name
    * tracking id
    * lot
    * state (is there any restriction around this?)
    * classifiedAs
    * image
    * note
  * The fields that the accounting must manage are:
    * resourceConformsTo (not really, by accounting, but set in stone by convention)
    * resourceQuantity
    * effortQuantity
    * primaryAccountable
    * currentLocation
    * stage (determined by the chain of inputOf and outputOf)
    * containedIn
* Once the customizations are done the data can be used to create the resources referenced by the event's resourceInventoriedAs and/or toResourceInventoriedAs.
* Relax constraints on resourceInventoriedAs and/or toResourceInventoriedAs in current accounting code.
* Potentially revise accounting code to make it fall more in line with the standard so it wont confuse people who look at it, or at least comment why it appears backwards (we took a shortcut instead of looking at the action.stateEffect because our data structure implied the correct logic, but it's confusing).

**Comment by adaburrows** [comment]
> No, we would never create a resourceInventoriedAs on a transfer. If it is not on the event, then we just don't know about it in our scope, and that is fine.

Oh, ok. I'll just have to add a conditional, since the accounting code we have assumes it's always there. I might also possibly have to rewrite the accounting code to remove a few assumptions about the effects. Like I said, the various parameters on the action are neglected and the data structures we used lets us assume a few things. I need to double check this now. This also means I might create a bit more feedback about the Valueflows data model.

> If by "debited" you mean incremented, yes.

Yes, looking at it with the goal of being able to synthesize double entry accounting, where debit is incrementing (as an asset) and credit is decrementing (as a liability).

> Also noting that a transferCustody will only affect the onhandQuantity

It was late and I was starting to fall asleep in my chair. The reality was starting slip away. I stopped wanting to type. :-) Thank you for clarifying. I was thinking the same way. All the accounting effects are laid out very nicely in your table and I'll probably rewrite the code to match it more completely. We have all the parameters for the actions, so we can specify the accounting exactly to your spec. I just didn't feel like typing out all the cases implied by your table last night in my comment, but I felt like I needed to write down something even if it was wrong or easily misunderstood.

> no restrictions, if there is a string in the event state, then put that in the resource state

Awesome, that makes it easy!

**Comment by adaburrows** [comment]
Ok, I just finished changing up the code to remove some of the assumptions. Also, I implemented to full updated Action data model, so we can use that to control how we show the fields in the Event entry form (with eventQuantity and createResource).

**Comment by adaburrows** [comment]
That actually did take longer than I expected, but this time I'm pretty sure about the form being done tomorrow now that everything feels in order. Then we can merge and clean up the repo.

**Comment by adaburrows** [comment]
@fosterlynn, I'm now starting to fully wrap my head around the accounting system and how the `Action.createResource` controls the logic of the accounting. I still need to add in some logic around that and I think we'll have a complete working version of that. However, I just realized we don't have a working definition of `optional` and `optionalTo`. Those individuals are referenced in the TTL file, but they are not defined: https://lab.allmende.io/valueflows/valueflows/-/blob/master/release-doc-in-process/all_vf.TTL#L1407

I thought I might have understood what `vf:optional` and `vf:optionalTo` are, but the more I think about it, the less well defined it feels. I think that might be intentional, but it also means that an operator can stop the flow by failing to inventory something. In fact, it gives the operator the ability to completely not inventory anything.

This also means that according to spec, I could give the operator the appropriate dropdown based on `vf:optional` or `vf:optionalTo`, and those drop downs would have a list of options of resources which `conformTo` the correct `ResourceSpecification`, an option to say it's N/A (where accounting doesn't happen for the event, and a UX element to add a new `EcnonomicResource` if desired.

However this also seems to be modulated by:
```For actions with the option decrementIncrement, the resourceInventoriedAs should be decremented (if there is one);and the toResourceInventoriedAs should be incremented (if there is one).```
Because in that case, it seems like those `optionalTo`s could have both of them present, but the the `optionalTo` definition implies they are both optional and the especially the `resourceInventoriedAs` is not even show on the screen.

This points to underspecification of the accounting state machine, or that `optionalTo` is not fully defined in that it allows the user to select:
* `resourceInventoriedAs` and `toResourceInventoriedAs` are both null
* `resourceInventoriedAs` is maybe null and `toResourceInventoriedAs` is present
* `resourceInventoriedAs` and `toResourceInventoriedAs` are both present

Or is there a restraint based on on the needing to be a `resourceInventoriedAs` needing to be present if the `Action` has one of the following has an `update` value and a restraint on their needing to be a `toResourceInventoriedAs` if one of the follow has an 'updateTo` value?
- accountingEffect
- onhandEffect
- locationEffect
- containedEffect
- accountableEffect
- stageEffect
- stateEffect

I think this is the last bit of clarification I need before I can finish this.

**Comment by fosterlynn** [comment]
>we don't have a working definition of optional and optionalTo.

Good catch, I'll fix in the ttl file.  And check the Actions page to clarify as needed.

In the meantime, these are values for createResource (as you saw), and mean:
* optional:  The app should give the user the option to create a new resource for resourceInventoriedAs
* optionalTo: The app should give the user the option to create a new resource for toResourceInventoriedAs

Otherwise, no new resource can be created by that action.

Also, to be more clear, the options in these cases are: create a new resource, update an existing resource, or neither.  The "neither" is because some agents won't inventory some resources, even those that theoretically could be inventoried, because it is not worth the trouble.  An example is things that they don't re-order based on inventory, but just do it by looking at things.  Like big bins of screws, or fluids.

>it also means that an operator can stop the flow by failing to inventory something

I suppose that could be true, depending on how the track/trace is done.  It could be based on ResourceSpecification in some cases.  Then something like the EU material passport could make use of it, because they mostly care what type of thing became part of a resource, or was used in its production, they're not tracking specific batches/lots, afaik.  And info from Bob: The FDA has some rules for those cases, and if they see a resource spec coming in on something they are tracing, they might look at inputs to that uninventoried resource just by type for some period of time in the past.  Which is what happens with an inventoried resource too, if it has been incremented multiple times but is not tracked by batch, or batches were combined into a new batch.

But that might not be in our doc on track/trace, I'll check.

>Because in that case, it seems like those optionalTos could have both of them present,

So, an event would never create 2 new resources, one for resourceInventoriedAs and one for toResourceInventoriedAs.  It is perhaps unfortunate we created this confusion for ourselves by adding the toResourceInventoriedAs after the IPO part of the model was pretty well set, using only resourceInventoriedAs.  I don't know if this is more confusing than the alternative, but quite possibly.

So to summarize, the optional and optionalTo are specifically tied to the resourceInventoriedAs and the toResourceInventoriedAs, and never imply the other.  But each of those individually are optional fields.

Hope this helps!

**Comment by fosterlynn** [comment]
Fixed in the turtle file.  Let me know if you see any problems with that, or further problems, thanks!

**Comment by adaburrows** [comment]
Adding in conversation between Lynn and I from elsewhere:

Jill:
> Hey, I'm sorry I confused you with my post. It's about both:
> * If we show the fields at all.
> * If we allow the option for adding a new resource.
> I'm assuming that for work and deliverService where there are no update or updateTo values in the Actions:
> * accountingEffect
> * onhandEffect
> * locationEffect
> * containedEffect
> * accountableEffect
> * stageEffect
> * stateEffect
> That no fields to select a resource should be shown at all and that should hold for any new action with the same patter.
> Also if one of those has an `update` value then only the ui for select resourceInventoriedAs is shown. If `updateTo` then both are shown.

Lynn: 
> > That no fields to select a resource should be shown at all and that should hold for any new action with the same patter.
>
> Yes because work and deliverService won't have resources.  Although I suppose someone might some day think they want them, but let's ignore that....
>
> > If updateTo then both are shown.
>
> No, just toResourceInventoriedAs 
> I think anything with a "To" in it, relates only the toResourceInventoriedAs

**Comment by adaburrows** [comment]
Then how do I transfer something with both a `resourceInventoriedAs` and a `toResourceInventoriedAs` when they differ?

It seems like this model is broken along that transfer process when both are required. It seems like the idea of having a process that mediates the transfer would be the way to go, and then we can get rid of the `toResourceInventoriedAs` completely from the spec. Is there a time we can play to chat about this? Or maybe I need to write out my thoughts more clearly so we can hash through them. Right now I need to go blow off some steam and won't get to any of the things I said I would today. 

Basically, when we want to transfer something, we need a reference to both its `resourceInventoriedAs` and `toResourceInventoriedAs`, because that's the only way we can know (without jumping up to the `Commitment` or `Intent` `resourceInventoriedAs`) which item we need to decrement in the decrementIncrement operations. Or we at least need to provide the option to have one.

I'm also looking at how to compute this in the simulation (where no `EconomicResources` are already created, but only created during the simulation) given only the `EconomicEvent` fields (`resourceConformsTo`, `provider`, `receiver`, `atLocation, `toLocation`) and the `EconomicResource` fields (`conformsTo`, `primaryAccountable`, `currentLocation`), given those fields, we can simulate what could happen without further user input.

```
/-------------------------\        /---------------------------\        /-------------------------\
| Resource A              |        | EconomicEvent             |        | Resource B              |
+-------------------------+        +---------------------------+        +-------------------------+
| conformsTo: R           |        | resourceConformsTo: R     |        | conformsTo: R           |
| currentLocation: AT     |        | atLocation: AT            |        | currentLocation: TO     |
| primaryAccountable: PRO |        | toLocation: TO            |        | primaryAccountable: REC |
\-------------------------/        | provider: PRO             |        \-------------------------/
                                   | receiver: REC             |                                   
                                   \---------------------------/                                   
```
First we make the above resources (which might not all exist if only the `toResourceInventoriedAs` was being used).

Then we can step through the accounting using the quantity adjustments on the `EconomicEvent` to adjust the resources involved in the transfer. In this case there is clearly a source `resourceInventoriedAs` that we need to decrement and there is a `toResourceInventoriedAs` the we need to increment.

That way of dealing with having basically no data entered from the user besides the connections in the graph, resourceSpecifications, agents, locations, and amounts works great. Except that when I do a transfer I don't simulate the `resourceInventoriedAs` not being there, because I can't know a priori that it won't be there. An excellent part of your model is that, typically, each increment/decrement action is split apart on each side of a process. That means in the simulation the other increment and decrement based actions work with the one specified resource. But in the case of a transfer, one of two things needs to happen:
* the entire quantity of an existing resource needs to be transferred
* one resource needs to be decremented and another (potentially just created by the user) needs to be incremented and it needs to happen in the same event because transfers don't have a process that splits them into two actions.

I see that taking transfers and splitting them into two phases and two actions would make this model whole again. It would potentially also allow simplifying part of that table since `decrementIncrement` would no longer be necessary. Instead there could be a transferOut and a transferIn that are pairsWith of each other and this whole logical contradiction in my mind goes away. 

Now in the case that I'm thinking through in my head, one resource needs to be decremented and another needs to be incremented, but the logic we've arrived at only lets them input the one new object and not the object that needs to be decremented. Assuming that we can use our `syntheticKey` we built for `EconomicResources`, we might find that multiple resources could be the input to the transfer, but the program cannot decide which one should be the `resourceInventoriedAs` when the transfer specifies only a `toResourceInventoriedAs`.

**Comment by adaburrows** [comment]
Or are we allowing that there are multiple transfer `EconomicEvent`s that might refer to the same happening? For instance, I transfer something from me to you and I only specify the `resourceInventoriedAs`. Then you also say I transferred something to you, but there is only a `toResourceInventoriedAs` in your event?

From a completeness point of view, I'd still rather see separate `transferOut` and `transferIn` actions that could be tied together by some shared object like a `Process` that could span membranes and be used to tie the whole network back together in bijective manner.

**Comment by fosterlynn** [comment]
I'm going to first answer in general, not specifically to the playspace, which might be a little different, depending on the use case.  I'm assuming you are writing general purpose logic for events affecting resources.

>Then how do I transfer something with both a resourceInventoriedAs and a toResourceInventoriedAs when they differ?

They will most often differ.  The easiest to think about are stock resources, the provider and receiver agents will each have their own stock of this kind of widget.  Maybe only some of many were involved in the event, which does not affect the identity of the resources, just the quantities.  They will generally be in different namespaces and have different identification schemes.  Serialized resources will keep the trackingIdentifier through a transfer, but usually not the instance identifiers as well as other properties.  Even a Van Gogh painting will probably be identified differently by the provider agent and the receiver agent.  The way to think about it is it the resource is part of accounting, not something that actually moves along in an event, which is not necessarily intuitive.

Moves (within the same agent) will differ much less often, usually only when location is part of the logical identifier. 

>It seems like the idea of having a process that mediates the transfer would be the way to go, and then we can get rid of the toResourceInventoriedAs completely from the spec.

That is another way to model it, and we have tried it both ways in software, and in VF discussions.  While having the same pattern was satisfying and in some ways cleaner, it was also in practice cumbersome, and basically we had 3 objects with mostly the same data.  (One dev who forked software that had that pattern - in this software a Transfer, with give and receive events - was so annoyed he rewrote all the transfer logic to have the simpler data model.) The current model also more matches what REA calls the "independent view", which VF takes, not the "trading partner view" or "enterprise view".  So I do think it could be done either way, and I'm not certain this was the right decision, but this is where we ended up.

One other angle, moves with 2 events and a mediating object would be even more redundant.

>First we make the above resources (which might not all exist if only the toResourceInventoriedAs was being used).

If a transfer doesn't have a resourceInventoriedAs, you can't create it.  It just isn't inventoried for the provider agent.  Or if the use case requires that, you'll have to require it on the event.  If all resources are inventoried, then a transfer must start with a resource that exists in the system, otherwise it doesn't make any sense, it had to get there somehow.

>Or are we allowing that there are multiple transfer EconomicEvents that might refer to the same happening?

Not in the same network.  If you are transferring something into the system, and the provider agent is not at all connected to the network or that software, they'll just decrement it on their side in their software, and we can't represent the complete independent view between those agents in one event.  If the agents are in different networks but using VF on holochain, I've assumed there is probably a way to make that happen without 2 events (or at least 2 differently identified events) representing the same thing, but I'm not close enough to how that could work in holochain.  It will be interesting.

Then, questions.  Does this have to be thought of differently in playspace?  It feels a bit different because the whole network feels like (and is) one namespace when you are using it to learn REA rather than to operate real processes and transfers with different independent agents involved.  But there are real networks that operate that way, such as Sensorica, which currently operates in one namespace, and atm one centralized app, with no projects (which are agents) having their own software.  And Code A might start with just one class of students, but want to expand to including the community in the processes and transfers. There will be a whole gradation of how independent the nodes are in networks that are represented in one app/dna.  Also, I'm unclear what role in this discussion the simulation plays, and what is the best way to understand what the requirement is for that?  (Not a sarcastic question, I just haven't thought about it since you showed it to us.)

**Comment by adaburrows** [comment]
> The current model also more matches what REA calls the "independent view", which VF takes, not the "trading partner view" or "enterprise view". So I do think it could be done either way, and I'm not certain this was the right decision, but this is where we ended up.

Yes, I understand this. This is actually how I've typically thought of resource flows and always thought the normal view of accounting was stupid for failing to account for the multiplicity of agents in the modeling. But I understood that those limited views ("dependent view") where a natural consequence of individuality and competition. The potential of Valueflows excited me, because it fit into my model of collectivism where the identity of the individual is merged into the cooperative survival of the collective super-organism.

> If a transfer doesn't have a resourceInventoriedAs, you can't create it. It just isn't inventoried for the provider agent. Or if the use case requires that, you'll have to require it on the event. If all resources are inventoried, then a transfer must start with a resource that exists in the system, otherwise it doesn't make any sense, it had to get there somehow.

Yep, I've been right there with you. I'm just fighting the contradiction between not allowing both fields to appear on the form when they both might be needed (this is based on a few things you said). I think I'm not explaining myself well enough. Let me separate the concerns, but first I'll answer more questions.

> Also, I'm unclear what role in this discussion the simulation plays, and what is the best way to understand what the requirement is for that? (Not a sarcastic question, I just haven't thought about it since you showed it to us.)

I was using it to understand what should happen to resources as the event dialog is being filled out. I modeled it ahead of time with those aspects because I thought there were certain possible inconsistencies between the logic being used to construct the form and what was needed for a particular event to actually happen and be accounted for. The code is the same as it is for accounting with a set of populated resources from user input. This is so that we can model what would happen given certain changes without actually committing it to hREA permanently until a user was certain what they were entering was correct. You could think of it like that validation warning on my original UX, you'd only save when you were sure it was right and valid. This meant needing to take a certain set of resources from hREA (or another store) and apply events to them to see if new Resources ever needed to be created (and therefore need a name or other parameter to be overridden by the user) in the path of events in the network. This detection was done through understanding that resources can technically be identified by the tuple of [conformsTo, primaryAccountable, currentLocation, trackingIdentifier, lot, containedIn, state, stage]. Of course, it is possible for multiple resources to have the same value of tuple, the only difference needs to be in the values omitted. The EconomicResources with these same values could be viewed as interchangeable for all intents and purposes. But we don't need to walk through that now. It's actually much less important than the rest of my issue.

### The setup

Here's my understanding prior to the last conversation. I hope to demonstrate the logical contradiction I'm encountering between the data model and the UX guidance I'm receiving.

#### Data model

There is set of EconomicResource objects conforming to the tuple containing the elements: (id, name, conformsTo, primaryAccountable, currentLocation, trackingIdentifier, lot, containedIn, state, stage, onhandQuantity, accountingQuantity, etc.). For the moment, let us assume that 

There is a set of EconomicEvent objects which conform to the tuple containing the elements: (id, action, provider, receiver, resourceInventoriedAs, toResourceInventoriedAs, inputOf, outputOf, atLocation, toLocation, resourceConformsTo, resourceQuantity, effortQuantity).

Each EconomicEvents specifies parameters to the accounting code that acts on **one or two** EconomicResources.

There is a set of Action objects which conform to the following tuple: (eventQuantity, inputOutput, accountingEffect, onhandEffect, pairsWith, locationEffect, containedEffect, accountableEffect, stageEffect, stateEffect)

Certain Action objects:
* have an eventQuantity value which indicate which of (resourceQuantity, effortQuantity) the EconomicEvent with the action is allowed to have.
* have inputOutput values indicating the can be an input to a Process or and output to a process.
* have no inputOutput values indicating they stand alone and constitute a whole action with accounting effects that occurs to one or two EconomicResources.
* have reciprocal pairsWith values. These have opposite inputOutput values and will always be inputs or outputs to a process.
* have a createResource which indicates when it is acceptable to create a new resource and which field it will use.
  * optional &mdash; optionally allow the creation of a new resource referenced by the resourceInventoriedAs field.
  * optionalTo &mdash; optionally allow the creation of a new resource referenced by the toResourceInventoriedAs field.

Each action object further specifies which mutations (and their potential arity) will be applied to the EconomicResource objects specified in the the accounting code.

The partial ordering of the EconomicEvent objects is given by the path order on the planning graphs they are connected to via both Processes and Satisfaction relationships. We are ignoring this detail for now and are assuming graph partial ordering is embedded in the total ordering of all EconomicEvents by the timestamp at which they were created.

When an EconomicEvent called E_t is accounted the referenced EconomicResource pair (resourceInventoriedAs, toResourceInventoriedAs) is adjusted based on the parameters in E_t and the Action, A_e, referenced by E_t. In most cases exactly one of these EconomicResources is references. If A_e has any `{*}Effect` fields, those mean an operation will be performed on the resource. An {accounting,onhand}Effect will adjust an {accounting,onhand}Quantity by the amount of E_t.resourceQuantity. The other effects will also perform adjustments on the EconomicResource.{*} fields, which correspond to the pattern {*}Effect on the Action.

The one Action where the arity of the mutator will include both of (resourceInventoriedAs, toResourceInventoriedAs) occurs when any of the {accounting,onhand}Effect values are set to decrementIncrement. If both resources (resourceInventoriedAs, toResourceInventoriedAs) are specified, the mutator will decrement resourceInventoriedAs and increment toResourceInventoriedAs. If only one of them is specified, only the specified EconomicResource will be adjusted (according to the pattern of decrement resourceInventoriedAs and increment toResourceInventoriedAs).

In our code, we actually enforce the arity of two and just pass in one of:
* (resourceInventoriedAs, toResourceInventoriedAs)
* (null, toResourceInventoriedAs)
* (resourceInventoriedAs, null)
This is equivalent for all intents and purposes.

Even though it has no direct effect on the accounting mutator, the Actions with the the decrementIncrement mutators all have a createResource specified as optionalTo. This may be used in the UI and is important to keep in mind in the following discussion.

#### UX logic

When users are creating EconomicEvents they must be able to select a subset of the EconomicResource pair (resourceInventoriedAs, toResourceInventoriedAs) and have the option to create new EconomicResources in certain cases (outlined above in the data model)

In most cases, the subset (resourceInventoriedAs) is all that is needed for an EconomicEvent to be properly accounted.

Let's examine the case of Actions with the createResource value of optionalTo, the accountingEffect and/or onhandEffect is decrementIncrement which indicates the arity of the mutator is allowed to be two, but could potentially be one.

If we look at what you said in response to my comment about figuring out how to display things based on the content of the Action object in the UI:

> >If updateTo then both are shown.
>
> No, just toResourceInventoriedAs
> I think anything with a "To" in it, relates only the toResourceInventoriedAs

There are a few test cases to look at (within the realm of the playspace and the possibilities it affords) within the the context of transferring an EconomicResource between two agents in the playspace, but I'll focus on one that leads to a direct logical contradiction with the above UX logic:

I want to transfer an EconomicResource from one agent in the playspace that calls it one thing to another agent in the playspace that calls it another thing:
* The two EconomicResources already exist from previous transactions.
* The two EconomicResources share the same conformsTo of "Widgets", because this OVN has settled on a standard.
* EconomicResource A
  * primaryAccountable "Xochitl's Fantastic Hardware Imporium".
  * name: "Cuetzpalin Keychain Design 2"
  * accountingQuantity: 70
  * onhandQuantity: 70
* EconomicResource B
  * primaryAccountable: "Bill and Ted's Excellent OVN".
  * name: "Sho-chee's Keychains"
  * accountingQuantity: 5
  * onhandQuantity: 5
So Xochitl is transferring 20 of A to B.

She goes to make the transfer and she can't because she can only select a (null, toResourceInventoriedAs) = (null, B), but she really needs to be able to select a (resourceInventoriedAs, toResourceInventoriedAs) = (A, B).

Even if she was only seeing her dependent view the correct accounting couldn't happen because in that case she'd need to select (resourceInventoriedAs, null) = (A, null) so that her inventory would be properly decremented.

**Comment by adaburrows** [comment]
We can also look at another example where we are transferring EconomicResources between two Agents A and B and some of the EconomicResource is not being transferred and Agent B doesn't have any EconomicResources conforming to the same ResourceSpecification.

A's EconomicResource:
* primaryAccountable: A
* name: "Handcrafted Midcentury Lawn Chair"
* accountingQuantity: 20
* onhandQuantity: 19
* currentLocation: "A's Woodworking Shop"

Let's assume they both have a Agreement with clauses that are Commitments corresponding to an transfer of money from B to A and a transfer of a "Handcrafted Midcentury Lawn Chair" from A to B.

A goes ahead and transfers 1 "Handcrafted Midcentury Lawn Chair" to B after receiving the money from the in workshop purchase A let's B load it up into their car. Both of them are using a tool that lets them track and trace their receipts and purchases and items they bought through Valueflows, so the POS system needs to create an EconomicEvent:
* it needs to specify the resourceInventoriedAs is A's "Handcrafted Midcentury Lawn Chair" so it can be credited (decremented, but some times I catch myself using old accounting terminology for assets)
* it needs to create a toResourceInventoriedAs, since this is B's first time buying "Handcrafted Midcentury Lawn Chair" from A:
  * primaryAccountable: B
  * name: "B's Handcrafted Midcentury Lawn Chair"
  * accountingQuantity: 1
  * onhandQuantity: 1
  * currentLocation: "B's Poolside Deck"

In the end A's EconomicResource should look like:
* primaryAccountable: A
* name: "Handcrafted Midcentury Lawn Chair"
* accountingQuantity: 19
* onhandQuantity: 18
* currentLocation: "A's Woodworking Shop"

If the POS was still behaving like our playspace (let's hope not!) A would have to manually create that and be sure that there was both a (resourceInventoriedAs, toResourceInventoriedAs). So the UI would need to show it.

**Comment by bhaugen** [comment]
I'm studying this issue and thread of dialog and will make one or more useful comments when I think I understand better, but I did want to say I really like "operator" instead of "user".

I will be focusing next on whether (and how) the receiving agent has agency, for example, can inventory a received resource or not as they want.

> That is another way to model it, and we have tried it both ways in software, and in VF discussions. While having the same pattern was satisfying and in some ways cleaner, it was also in practice cumbersome, and basically we had 3 objects with mostly the same data. (One dev who forked software that had that pattern - in this software a Transfer, with give and receive events - was so annoyed he rewrote all the transfer logic to have the simpler data model.)

Seems possible that the other dev was wrong. We called the two events Give and Take, where the provider executed the Give and the receiver executed the Take and thus both of them had full agency. That might look and feel different in a monolithic system (as the other dev was working in) vs a distributed system.

That's just a drive-by comment for now, not a proposal that two  events should be operated now, but I'll be studying the details from the viewpoint of two independent agents in a transfer.

**Comment by fosterlynn** [comment]
One simple thing before the complicated thing. :)  Re. `updateTo`, that is only used for location, primaryAccountable, state, and stage.  It isn't meant to have update meaning on anything else.  I'm not sure that is where it was confusing, but just in case.  It isn't meant to have anything to do with what is shown to the operator in terms of the 1 or 2 resources.  (I'm going to try to remember "operator", I like it too!)  If we need to work with naming or logic or anything else on those new Action behaviors, either the UI or the mutation related ones, that will be easy to do at this stage of people's usage of the vocab, the new ones are just this project so far afaik. And the old ones may be fair game without bothering anyone too much either, but I would need to check with people.  

Complicated thing next but that will take some thought, will try to do it in the next few hours, mostly to try to lay out the trade offs, and Bob's comment is very useful, also the agency problem shows up in your POS example....

**Comment by adaburrows** [comment]
> I'm the operator with my pocket calculator.
- Kraftwerk [https://youtu.be/eSBybJGZoCU]

**Comment by adaburrows** [comment]
> Re. updateTo, that is only used for location, primaryAccountable, state, and stage. It isn't meant to have update meaning on anything else. I'm not sure that is where it was confusing, but just in case.

If I mentioned that as the filed upon which any of the UI hinged, it was a typo. All of the references were supposed to be ones with optionalTo and with decrementIncrement {accounting,onhand}Effect where it is possible to put both (resourceInventoriedAs, toResourceInventoriedAs) and the UI should support that in the case that we all prioritizing the independent view in the playspace.

**Comment by adaburrows** [comment]
> Seems possible that the other dev was wrong. We called the two events Give and Take, where the provider executed the Give and the receiver executed the Take and thus both of them had full agency. That might look and feel different in a monolithic system (as the other dev was working in) vs a distributed system.

I wouldn't make a judgement about right or wrong, I can see that devs perspective and I can see my mathematical perspective and I can't decide between them just yet. I'm thinking about how this works for something that might work in the playspace where everything is in the same membrane, but also how it might work across multiple hREA instances or multiple instances of any other Valueflows implementation (or across both).

However, the core of this problem is really just getting to the point where this is well defined: There are multiple use cases of a transfer Action that may occur:
* (resourceInventoriedAs, null) - Corresponds to "give/transferOut" Actions in the alternate models discussed above. Maps onto the dependent view, where only certain things are known, but not everything in the network.
* (null, toResourceInventoriedAs) - Corresponds to "take/transferIn" Actions in the alternate models above. Maps onto the dependent view, where only certain things are known, but not everything in the network.
* (resourceInventoriedAs, toResourceInventoriedAs) - A transfer where all knowledge of inventory is shared, there is no membrane. Maps onto the independent view, where everything is defined in the network and fully discoverable.

In the case where things fracture into local networks with some kind of membrane preventing access to all of the other network's data, it would be nice to have a shared object that could tie together two of the partial EconomicEffects. In Holochain parlance, there could be a lobby that contains these references. I feel like for now we could just use a Process as this placeholder. This would allow doing tracking and tracing across membranes. Currently, I believe hREA gets around this sort of thing by duplicating objects.

**Comment by fosterlynn** [comment]
>(resourceInventoriedAs, toResourceInventoriedAs) - A transfer where all knowledge of inventory is shared, there is no membrane. Maps onto the independent view, where everything is defined in the network and fully discoverable.

One reminder, that even inside a network with no membranes, sometimes agents don't inventory everything, it is not worth the trouble.
